### ISSUE TITLE
Bug in WCS slicing with negative step

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#4529]
In the following example, I should get the same world coordinates for the velocity axis regardless of whether I reverse the order of the WCS or not:

``` python
In [1]: import numpy as np

In [2]: from astropy.wcs import WCS

In [3]: from astropy.io import fits

In [4]: from astropy.utils.data import get_pkg_data_filename

In [5]: filename = get_pkg_data_filename('l1448/l1448_13co.fits')

In [6]: header = fits.getheader(filename)

In [7]: wcs = WCS(filename)

In [8]: x = np.repeat(0, header['NAXIS3'])

In [9]: y = np.repeat(0, header['NAXIS3'])

In [10]: z = np.arange(header['NAXIS3'])

In [11]: wx, wy, wz = wcs.wcs_pix2world(x, y, z, 0)

In [12]: print(wz)
[ 2528.19489695  2594.61850695  2661.04211695  2727.46572695  2793.88933695
  2860.31294695  2926.73655695  2993.16016695  3059.58377695  3126.00738695
  3192.43099695  3258.85460695  3325.27821695  3391.70182695  3458.12543695
  3524.54904695  3590.97265695  3657.39626695  3723.81987695  3790.24348695
  3856.66709695  3923.09070695  3989.51431695  4055.93792695  4122.36153695
  4188.78514695  4255.20875695  4321.63236695  4388.05597695  4454.47958695
  4520.90319695  4587.32680695  4653.75041695  4720.17402695  4786.59763695
  4853.02124695  4919.44485695  4985.86846695  5052.29207695  5118.71568695
  5185.13929695  5251.56290695  5317.98651695  5384.41012695  5450.83373695
  5517.25734695  5583.68095695  5650.10456695  5716.52817695  5782.95178695
  5849.37539695  5915.79900695  5982.22261695]

In [13]: wcs_r = wcs[::-1]

In [14]: wx, wy, wz = wcs_r.wcs_pix2world(x, y, z, 0)

In [15]: print(wz[::-1])
[ -992.25643305  -925.83282305  -859.40921305  -792.98560305  -726.56199305
  -660.13838305  -593.71477305  -527.29116305  -460.86755305  -394.44394305
  -328.02033305  -261.59672305  -195.17311305  -128.74950305   -62.32589305
     4.09771695    70.52132695   136.94493695   203.36854695   269.79215695
   336.21576695   402.63937695   469.06298695   535.48659695   601.91020695
   668.33381695   734.75742695   801.18103695   867.60464695   934.02825695
  1000.45186695  1066.87547695  1133.29908695  1199.72269695  1266.14630695
  1332.56991695  1398.99352695  1465.41713695  1531.84074695  1598.26435695
  1664.68796695  1731.11157695  1797.53518695  1863.95879695  1930.38240695
  1996.80601695  2063.22962695  2129.65323695  2196.07684695  2262.50045695
  2328.92406695  2395.34767695  2461.77128695]
```

cc @nden @keflavich

**Comment by astrofrog** [comment]
Here is a simpler example that illustrates the issue:

```
In [28]: wcs = WCS(naxis=1)

In [29]: wcs.wcs.crval = [5]

In [30]: wcs.wcs.crpix = [4]

In [31]: wcs.wcs.cdelt = [2]

In [32]: xp = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [33]: xw1 = wcs.wcs_pix2world(xp, 0)

In [34]: xw2 = wcs[::-1].wcs_pix2world(xp, 0)[::-1]

In [35]: xw1
Out[35]: [array([ -1.,   1.,   3.,   5.,   7.,   9.,  11.,  13.,  15.,  17.])]

In [36]: xw2
Out[36]: [array([ -3.,  -5.,  -7.,  -9., -11., -13., -15., -17., -19., -21.])]
```

The issue is that ``crpix`` should be updated to ``NAXIS1 - 1 - CRPIX1`` in this case. However, WCS objects don't properly contain a reference to the dataset they apply to, so ``NAXIS1`` isn't known (I know we *actually* keep the naxis values internally but these are not necessarily guaranteed to be defined.

One solution would be that if the step is negative, the slice **has** to contain the end value, so for instance ``wcs[:10:-1]``.

@nden @keflavich - what do you think?

**Comment by astrofrog** [comment]
Hmm, but actually in Numpy notation, the start/end are very confusing when using slicing with a negative step:

```
In [2]: x = np.arange(10)

In [3]: x[:5:-1]
Out[3]: array([9, 8, 7, 6])
```

so maybe not an ideal solution.

**Comment by astrofrog** [comment]
Maybe a better solution is to add an optional ``shape=`` to ``WCS.slice`` and if someone does ``[::-1]`` we tell them to use ``.slice`` instead.

**Comment by nden** [comment]
@astrofrog I am a bit confused about this, correct me if I'm wrong. The slicing of a WCS works by changing the WCS, not by slicing the result. Is this correct?
If so, should a negative slice change the order of the axes as well?

**Comment by astrofrog** [comment]
@nden - if you have a 2D WCS associated with a 2D image ``array``, then doing e.g. ``wcs[slice]`` will adjust the WCS so that it would be correct on the sliced array ``array[slice]``. So doing e.g. ``wcs[::-1,::-1]`` would invert the order of each individual axes but not swap the axes. However, the issue is that if you want to change the order of the axis, then if CRPIX1 was 1, it should be changed to NAXIS1. But we can't do that without knowledge of the image size.

**Comment by astrofrog** [comment]
@nden - and indeed it changes the WCS itself, it doesn't adjust the result.

**Comment by keflavich** [comment]
@astrofrog agreed that the end value specification is confusing for this.  I am in favor of disallowing `[::-1]` and raising an error message saying `Use this.slice(slice(None, None, -1), shape=(naxis1,naxis2)) to invert an axis.  Inverting a FITS axis requires specifying a new CRPIX referenced to the other end of the image.  See issue 4592`, or something similarly extensive & detailed.

**Comment by keflavich** [comment]
see https://github.com/keflavich/astropy/commit/a07c4af56b101bcb4155771b6a1224fe54d8d53b for a less complete solution

**Comment by keflavich** [comment]
I think this should have been closed by #4962
